rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+- Every hiring manager loves javascript cause it's easy to hire javascript developers. The main problem with this approach is there are so many kinds of javascript development and ways to write javascript that hiring a javascript developer is not a gaurantee they'll be able to write javascript the way your team uses it.
+
+#### Some Notes On Our Javascript
+
+- Thankfully this team has unified on a coding style, linter, tool chain, package manager, and approach to typing. +1 on that.
+- We have not however documented why we made the choices we did. We should do so ASAP. In particular types in JSDoc style comments and why we do it needs documentation. This includes our sporadic dropping into Typescript and why we do it. (For example: 'why is there an api.js and an api.ts and the js file just says `export {}` -- the answer is 'cause modules' but is EXTREMELY not clear)
+- The w3up repo is huge, and honestly, even if you know javascript, there's a lot to get up to speed on. We should have a guide to the code.
+
+### Go
+
+- Go is a highly performant garbage collected language like Java, but is much simpler than Java
+- Go has an extremely powerful concurrency model with green threads
+- These two components together make Go one of the most powerful languages for writing servers of any kind. The challenge of server programming is not super fast compute, but super high concurrency execution. Go excels at this, which matters cause our project has a lot of servers involved.
+- Go's ability to do binary distributions by default is a major plus for kind of project.
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
+- Lots and lots of Filecoin tooling and software is written in Go. In fact, W3S already has Go in it, we just don't maintain it -- it's called Spade
+- Garbage collection and green threads present challenges for using Go over an FFI, or for compiling go to WebAssembly, though TinyGo has improved this story by reducing the runtime size.
+
+### Rust
+
+Caveat from Hannah: I am not a highly experienced Rust dev
+
+- Rust is a better C/C++. Like definitely much better. This is huge cause no one wrote a better C/C++ for low level programming for 20 years.
+- However, by and large, you should probably not write programs in Rust unless you'd consider writing them in C/C++. There is minimal evidence one can be as productive in Rust as higher level languages.
+- In particular, you should probably not write web servers in Rust or any other asynchronous programs unless you have extra time on your hands which we don't.
```

Comment:
I find some of the statements about Rust inaccurate. I find myself a lot more productive reading and understanding Rust code than I do with go as it's type system forces you into specific design patterns. All the PL go code I had to deal is a nightmare in comparison, no comments terrible naming (one or two letter variable names 😅).

Reply:
I think this is a great article describing why discord has switched from go to rust https://discord.com/blog/why-discord-is-switching-from-go-to-rust

Reply:
That was an interesting read. Thanks 👍

Reply:
This is great. Also, the async library has come a long way, and its progress is very promising.

Reply:
I think this is a good counter-example... but is also a lesson when porting from javascript to Go/Rust/etc. might be a bad idea. https://www.youtube.com/watch?v=_SzvJJ3_6M0

The gist is that a huge amount of time and effort was spent to arrive at an outcome with much more code that offered no significant advantages.

Reply:
@Gozala 

The Discord example is an argument that when a code path becomes super hot, you can get more optimization by switching from a garbage collected language to one with manual memory management.

This feels both correct and also somewhat trivially obvious. Again this is where my "consider rust when you'd consider C/C++" statement comes from.

The more interesting question is when does a code path become hot enough to merit this sort of optimization, and what is the cost of trying to do this for all code paths, even the ones that don't need it.